### ISSUE TITLE
ActiveRecord model instance's singleton class raises exception when 'inspect' is called on it

### ISSUE THREAD

**Issue by maxfelsher** [rails/rails#53013]
### Steps to reproduce
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  # gem "rails"
  
  gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_singleton_of_new_model_instance
    Post.new.singleton_class.inspect
    pass
  end

  def test_singleton_of_existing_model_instance
    Post.create!
    Post.first.singleton_class.inspect
    pass
  end
end
```

### Expected behavior
No errors. Tests pass.

### Actual behavior
Exceptions raised.

```
  1) Error:
BugTest#test_singleton_of_new_model_instance:
NoMethodError: undefined method `table_name' for nil
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:620:in `compute_table_name'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:299:in `reset_table_name'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:261:in `table_name'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:418:in `table_exists?'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/core.rb:359:in `inspect'
    singleton_inspect.rb:33:in `test_singleton_of_new_model_instance'

  2) Error:
BugTest#test_singleton_of_existing_model_instance:
NoMethodError: undefined method `table_name' for nil
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:620:in `compute_table_name'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:299:in `reset_table_name'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:261:in `table_name'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/model_schema.rb:418:in `table_exists?'
    /usr/local/rvm/gems/ruby-3.3.4/bundler/gems/rails-9c41f5828cd4/activerecord/lib/active_record/core.rb:359:in `inspect'
    singleton_inspect.rb:39:in `test_singleton_of_existing_model_instance'

2 runs, 0 assertions, 0 failures, 2 errors, 0 skips
```

### System configuration
**Rails version**: This was tested against the `main` branch. Modifying the executable test case to use earlier versions of Rails leads to an exception being raised in the tests at least as far back as Rails 6.0. I tried going further back, but I ran into issues that didn't seem worth working through (the setup code from the bug report template started to error with Rails 5.2 and the version of Ruby I was using).

**Ruby version**: 3.3.4 (though likely it affects other versions as well)

### More information
I noticed this when I was trying to debug an Active Record model and I called `some_model_instance.singleton_class.ancestors` in a Rails console. I'm not sure it would have been that helpful anyway, but I was surprised that it errored out.

The issue seems to be caused by the ActiveRecord model class' custom `inspect` method, which is inherited by the singleton class:
https://github.com/rails/rails/blob/31818edb035266c487510db933ed97a33093751e/activerecord/lib/active_record/core.rb#L351-L365

The first three conditionals evaluate to false, so the method tries to call `table_exists?`, which eventually calls `table_name` on a nil `base_class`. I didn't look too closely into why `base_class` is nil in this case, but I'm wondering if a singleton class never triggers `inherited` or any of the other attempts to make sure the base class is set.

Personally, I would be fine with falling back to the super-method's `to_s` behavior if `singleton_class?` is true, assuming there are no performance concerns.

Basically the same problem was raised a few years ago in #38293, but that issue was automatically closed after going stale.
